packages/react/src/hooks/useSequentialRoomConnectDisconnect.ts

```diff
@@ -0,0 +1,126 @@
+import { Mutex, Room } from 'livekit-client';
+import { useCallback, useMemo, useRef } from 'react';
+import { log } from '@livekit/components-core';
+
+const CONNECT_DISCONNECT_WARNING_THRESHOLD_MS = 400;
+const CONNECT_DISCONNECT_WARNING_THRESHOLD_QUANTITY = 2;
+
+/** @public */
+export type UseSequentialRoomConnectDisconnectResults = {
+  connect: typeof Room.prototype.connect;
+  disconnect: typeof Room.prototype.disconnect;
+};
+
+/**
+ * When calling room.disconnect() as part of a React useEffect cleanup function, it is possible for
+ * a room.connect(...) in the effect body to start running while the room.disconnect() is still
+ * running. This hook sequentializes these two operations, so they always happen in order and
+ * never overlap.
+ *
+ * @example
+ * ```ts
+ * const { connect, disconnect } = useSequentialRoomConnectDisconnect(room);
+ *
+ * // Connecting to a room:
+ * useEffect(() => {
+ *   connect();
+ *   return () => disconnect();
+ * }, [connect, disconnect]);
+ * ```
+ *
+ * @public
+ */
+export function useSequentialRoomConnectDisconnect(
+  room: Room,
```

Comment:
It might be good to for room to have some sort of inactive value (ie, `null` / `undefined`) so if a user did create a room later they didn't have to play games to conditionally run this hook.

Reply:
Done